### PR TITLE
kvm: use libvirt bindings instead of virsh command, and add more actions

### PR DESCRIPTION
**Pull request by mie00**
see #88 and #108

### CODE REVIEW

File: core0/kvm/manager.go [hunk 1]
```diff
@@ -410,131 +511,420 @@ func (m *kvmManager) create(cmd *core.Command) (interface{}, error) {
...
 	}
+	return struct {
+		UUID string `json:"uuid"`
+	}{uuid}, nil
+}
 
-	tmp.Close()
+type DomainActionParams struct {
+	UUID string `json:"uuid"`
+}
```
Comment (grimpy) [inline core0/kvm/manager.go]: All types are defined on top of the file. Also if you rename this to DomainUUID or something like that you could reuse it in line 535

File: core0/kvm/manager.go [hunk 2]
```diff
@@ -410,131 +511,420 @@ func (m *kvmManager) create(cmd *core.Command) (interface{}, error) {
...
+		return nil, fmt.Errorf("couldn't find domain with the uuid %s", params.UUID)
+	}
+	if err := domain.Destroy(); err != nil {
+		return nil, fmt.Errorf("failed to destroy machine: %s", err)
 	}
+	name, err := domain.GetName()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get machine name with the uuid %s", params.UUID)
+	}
+	m.unPortForward(name)
```
Comment (grimpy) [inline core0/kvm/manager.go]: everything is refactored to use uuid probably this method should be aswell as name is not guaranteed to be unique

File: core0/kvm/manager.go [hunk 3]
```diff
@@ -410,131 +511,420 @@ func (m *kvmManager) create(cmd *core.Command) (interface{}, error) {
...
+	conn, err := libvirt.NewConnect("qemu:///system")
+	if err != nil {
+		return nil, fmt.Errorf("failed to start a qemu connection: %s", err)
+	}
+	defer conn.Close()
+
+	domain, err := conn.LookupDomainByUUIDString(params.UUID)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't find domain with the uuid %s", params.UUID)
+	}
```
Comment (grimpy) [inline core0/kvm/manager.go]: I've seen the above block repeated 3 times now, maybe its worth to add a method to get the domain object

File: pyclient/g8core/client.py [hunk 4]
```diff
@@ -1312,6 +1565,7 @@ def __init__(self, host, port=6379, password="", db=0, timeout=None):
         self._btrfs_manager = BtrfsManager(self)
         self._zerotier = ZerotierManager(self)
         self._experimntal = Experimental(self)
+        self._kvm = KvmManager(self)
```
Comment (grimpy) [inline pyclient/g8core/client.py]: kvm was added directly under client which is ok, but please remove it from experimental

File: core0/kvm/manager.go (hunk 3 above)
Comment (mie00) [inline core0/kvm/manager.go]: The idea is, even we did that, we will have to return domain, err from inside the function, and its usage will be exactly the same

Comment (grimpy) [inline core0/kvm/manager.go]: i meant the entire block incyuding the domainactions params parse till getting the domain
those are 3 errors you check

if you have a method
you will only have to check on the error once

**Pull request merged**